File: Java/src/_0483_Smallest_Good_Base/Solution.java
package _0483_Smallest_Good_Base;

/**
 * Date: 2021/6/18 17:17
 * 执行用时：5 ms, 在所有 Java 提交中击败了14.10%的用户
 * 内存消耗：36.6 MB, 在所有 Java 提交中击败了66.67%的用户
 */
public class Solution {

    public String smallestGoodBase(String n) {
        long N = Long.parseLong(n);
        for (int t = (int) (Math.log(N) / Math.log(2) + 1); t >= 3; t--) {
            long k = (long) Math.pow(N, 1.0 / (t - 1));
            long r = 0;
            for (int i = 0; i < t; i++) r = r * k + 1;
            if (r == N) return k + "";
        }
        return String.valueOf(N - 1);
    }

    public static void main(String[] args) {

        System.out.println((new Solution()).smallestGoodBase("13"));  // 3
        System.out.println((new Solution()).smallestGoodBase("1000000000000000000"));  // 999999999999999999
    }
}

File: Python/_0483_Smallest_Good_Base/Solution.py
# coding=utf-8
# Date: 2021/6/18 17:35
import math


# 执行用时：44 ms, 在所有 Python3 提交中击败了90.00%的用户
# 内存消耗：14.9 MB, 在所有 Python3 提交中击败了20.00%的用户
class Solution:
    def smallestGoodBase(self, n: str) -> str:
        N = int(n)
        t = int(math.log2(N) + 1)
        while t >= 3:
            k = int(pow(N, 1.0 / (t - 1)))
            r = 0
            for i in range(t):
                r = r * k + 1
            if r == N:
                return str(k)
            t -= 1
        return str(N - 1)


if __name__ == "__main__":
    print(Solution().smallestGoodBase("13"))  # 3
